Refactor metadata model

lukpueh opened the pull request:
This is another attempt to fix the implementation-specification `has-a` vs. `is-a` metadata format mismatch as documented in #98.

Santiago's solution for this issue, turned out to be slightly over-engineered, which in all fairness he already hinted at in the PR's description (see #120). This PR simplifies the metadata model by removing the adapter pattern and custom `__getattr__` implementation. 

What remains is a container class for signed metadata (`Metablock`) with a `signed` and a `signatures` attribute and methods for JSON serialization and deserialization, reading and writing to disk, and creating and verifying signatures.
The `signed` attribute of `Metablock` holds either `Link` or `Layout` objects, which contain the actual metadata information. The `signatures` attribute holds the cryptographic signatures computed over the canonical json representation of the `Link` or `Layout` contained in the `signed` attribute.

It wasn't feasible to create small, self-contained commits that leave the system in a stable state. However, the commit messages can be used to incrementally understand the modifications.

I also used the chance for an overall review of the system and to make small updates and fixes, especially regarding comments and docstrings in various places.

Commit:
Move Metablock class to own module

This commit does not change any logic, it simply moves the
Metablock class from the common module to a new metadata module
and updates the affected import statements.

In the future Metablock won't be an abstract superclass for Link
or Layout (adapters) but instead will be a concrete metadata
class that contains instances of Link or Layout.

For clarity and to avoid problems with circular imports (Metablock
will require Link and Layout, and Link and Layout will require
ValidationMixin from the common module, ...) this commit moves the
Metablock class from the common module to the metadata module.

Commit:
Add metadata loading method to Metablock (WIP)

Note: This is part of a series of refactoring commits.

Adds static method to Metablock class to load signed Metadata
that contains Link or Layout data from disk. This method will
replace Link's and Layout's `read_from_disk` static methods.

The commit also removes those static methods from the Link and
Layout class.

Commit:
Move and update link loading method (WIP)

Note: This is part of a series of refactoring commits.

This commit moves the function that loads all the links as defined
in a layout from a static method of the Layout class to an
autonomous function in verifylib and updates the function to
fit the WIP adapter refactoring.

Having this function as class method in Layout does not seem
intuitive anymore, the function should rather be seen as part of
the whole final product verification process. This commit also
helps if we we want to further reduce OOP concepts in the future.

Commit:
Remove Link and Layout dump methods (WIP)

Note: This is part of a series of refactoring commits.

This commit removes the methods to write a Link or Layout object
to disk.

In the future Link and Layout objects should only be written to
and read from disk when contained in a Metablock object
(add signatures). Hence, Metablock provides the loading and
dumping methods.

Commit:
Move read Link/Layout-from-dict (WIP)

Note: This is part of a series of refactoring commits.

This commit moves the static read methods that are used to parse
a dictionary in order to create a Link or Layout object to a
place where they will be preserved, since in the course of this
refactor the adapter classes will be removed.

Commit:
Purge adaptor pattern (WIP)

Note: This is part of a series of refactoring commits.

This commit fully removes the adaptor pattern used to model the
relationship between generic signed Metadata and Link and Layout
objects.

The pattern was introduced to facilitate the transation
from *Link or Layout `is-a` signed metadata* to *signed metadata
`has-a` Link or layout* relationship cf.i in-toto/in-toto#120

Unfortunately, the pattern added a lot of extra complexity, which
made the code harder to understand and as a consequence introduced
bugs.

The commit removes all adaptor related logic from Metadata, Layout
and Link classes.

Furthermore it removes the adaptor classes Link and Layout and
renames the data holding classes LinkSignable and
LayoutSignable back to Link and Layout respectively.

Commit:
Add _type property to Metablock (WIP)

Note: This is part of a series of refactoring commits.

Adds a property to Metablock to get the type of the contained
`signed` object, i.e. "link" or "layout".

Commit:
Add short filename format to link module (WIP)

Note: This is part of a series of refactoring commits.

The link method that was used to write link objects to disk was
removed in favor of a Metablock dump method, which always
requires the destination path to be passed as argument.

Commit:
Adopt refactoring changes in in-toto-sign (WIP)

Note: This is part of a series of refactoring commits.

- Replaces manual metadata loading with call to Metablock.load
- Updates extracting the name property from the signed part of
  a metadata object

Commit:
Adopt refactoring changes in in-toto-verify (WIP)

Note: This is part of a series of refactoring commits.

- Replace Layout.read_from_file with Metablock.load

Commit:
Adopt refactoring changes in runlib and verifylib (WIP)

Note: This is part of a series of refactoring commits.

Adopts loading and dumping metdata, and accessing Link and Layout
properties inside a Metablock object

Commit:
Adopt refactoring changes in unittests

Note: This is part of a series of refactoring commits.

Adopts loading and dumping metdata and accessing Link and Layout
properties inside a Metablock object.

Commit:
Fix various imports

Note: This is part of a series of refactoring commits.

Add missing imports, remove obsolete imports, remove remaining
occurrences of double underscore import naming scheme.

Commit:
Fix various docstrings

Note: This is part of a series of refactoring commits.

Updates/fixes header and function docstrings in all modules
affected by the refactor.

lukpueh commented on in_toto/models/metadata.py:
```diff
@@ -0,0 +1,207 @@
+    # We are mixing 3 JSON string formats here: The value of "signed" is
+    # "pretty printed canonical json", the value of "signatures" is
+    # "canonical json" and the container is just "json".
+    # Is this really what we want?
```
@SantiagoTorres, what do you think? I'd say, either we make it all (pretty printed) canonical JSON or none of it.
The only really important thing is that the JSON string representation of the object held in the `signed` attribute (Link or Layout) is deterministic when we create or verify signatures.

awwad replied:
Remember to make this an issue for later.

lukpueh replied:
https://github.com/in-toto/in-toto/issues/144

awwad commented on in_toto/runlib.py:
```diff
@@ -31,12 +31,15 @@
 from in_toto import log
-from in_toto.models.link import (UNFINISHED_FILENAME_FORMAT, FILENAME_FORMAT)
+from in_toto.models.link import (UNFINISHED_FILENAME_FORMAT, FILENAME_FORMAT,
+    FILENAME_FORMAT_SHORT)
```
Remember that when you `from ... import` a string, it is copied from the source module at import time. Thus, if FILENAME_FORMAT is changed in any module after this module is imported the first time, the value in this module and the value in `in_toto.models.link` (or any module that uses a normal import of `in_toto.models.link`) will not match. Testing is one context in which such a pattern might be used, as would be any kind of initial configuration setup at a higher level than this code. (This is true not just of strings, but of anything immutable or any variable you reassign later.)

It doesn't necessarily mean you shouldn't do this, but it's worth keeping in mind. (I've been burned by this.)

lukpueh replied:
I'll fix this with https://github.com/in-toto/in-toto/issues/145

awwad commented on in_toto/models/layout.py:
```diff
@@ -222,6 +101,25 @@ def __init__(self, **kwargs):
+  @staticmethod
+  def read(data):
+    """Static method to instantiate a new Layout from a Python dictionary """
+    signatures = data.get('signatures', [])
```
The signatures line here looks like it's not used - which is good, since it seems like it's none of class Layout's business. So this line was meant to be deleted and was missed, I think?

awwad commented on in_toto/models/layout.py:
```diff
@@ -222,6 +101,25 @@ def __init__(self, **kwargs):
+
+  @staticmethod
+  def read(data):
+    """Static method to instantiate a new Layout from a Python dictionary """
```
I think it would be good to have a bit more docstring here explaining what data looks like.

Is it like this?
`{"signed": {"steps": ..., "inspect": ..., ...}, "signatures": [...]}`

Or like this?
`{"steps": ..., "inspect": ..., ... }`

The latter, right?

awwad commented on in_toto/models/layout.py:
```diff
@@ -294,7 +192,7 @@ def _validate_steps_and_inspections(self):
       names_seen.add(inspection.name)
 
 @attr.s(repr=False, init=False)
```
Recognizing and looking into [attrs](http://www.attrs.org/en/stable/api.html) took a little while. I like it, but it's probably worth adding an inline comment. What do you think? Something like `# automatically provide common class methods like __eq__()` 

Also, I admit that it took me a moment to not parse "attr.s" as "attr.string". (use of "serious business" alias attr.attrs might be better).

lukpueh replied:
I think I'll leave it be for now. But I'll keep your comments regarding `attr` in the back of my mind.

awwad commented on in_toto/models/layout.py:
```diff
@@ -294,7 +192,7 @@ def _validate_steps_and_inspections(self):
 
 @attr.s(repr=False, init=False)
-class Step(models__common.ValidationMixin):
+class Step(ValidationMixin):
```
I thought about whether or not this being the only mix-in - and the only class that Step, Inspection, and Signable inherit from - meant that it would be more readable to just make ValidationMixin into a generic in-toto class parent class, something like `InTotoObject`... but I dunno.

The kicker, though, is that the entire ValidationMixin class is four lines (define, for, if, and one function call), and only used in three classes, so maybe you should just put validate() in each class and save the reader from having to contemplate any subclassing or mix-ins or bouncing back and forth if they're reading the class or looking for a validate() method.

In fact, even class Signable only has two lines and is only inherited by two classes, so you could strip **all** class inheritance from in-toto and it would cost you a few of lines of code duplication but make the code more readable and less bouncy and entirely get rid of module `common.py`. (You might end up with about the same number of lines after removal of super calls and imports and such.)

Just a thought.

lukpueh replied:
I had the exact same thoughts, but concluded that the PR was already big enough. However, in the grand scheme of things we'll probably remove OOP altogether. But for now I think the inheritance cases are easy enough to grasp and the alternative of having duplicate code is not ideal either.

awwad commented on in_toto/models/common.py:
```diff
@@ -15,140 +15,44 @@
+
 
 @attr.s(repr=False, init=False)
 class Signable(ValidationMixin):
```
This doesn't directly relate to this PR, but I'm observing it at this code state and this PR does touch data representation, so:

Since `__repr__` is cryptographically important for class `Signable`, you should add a clear note to `Signable.__repr__()` that wards off minor tinkering. (Something like `# Please note that the returned string from this function is the content signed by ... so changes affect what data is signed, break backward compatibility with old data, and any example data`)

awwad reviewed (commented):
Looks good to me. Feedback consists of architectural thoughts and some minor edits.

JustinCappos commented:
On Thu, Nov 2, 2017 at 3:14 PM, Sebastien Awwad <notifications@github.com>
wrote:

> *@awwad* commented on this pull request.
>
> Looks good to me. Feedback consists of architectural thoughts and some
> minor edits.
> ------------------------------
>
> In in_toto/runlib.py
> <https://github.com/in-toto/in-toto/pull/142#discussion_r148342777>:
>
> > @@ -31,12 +31,15 @@
>
>  import in_toto.settings
>  from in_toto import log
> -from in_toto.models.link import (UNFINISHED_FILENAME_FORMAT, FILENAME_FORMAT)
> +from in_toto.models.link import (UNFINISHED_FILENAME_FORMAT, FILENAME_FORMAT,
> +    FILENAME_FORMAT_SHORT)
>
> Remember that when you from ... import a string, it is copied from the
> source module at import time. Thus, if FILENAME_FORMAT is changed in any
> module after this module is imported the first time, the value in this
> module and the value in in_toto.models.link (or any module that uses a
> normal import of in_toto.models.link) will not match. Testing is one
> context in which such a pattern might be used, as would be any kind of
> initial configuration setup at a higher level than this code. (This is true
> not just of strings, but of anything immutable or any variable you reassign
> later.)
>
> It doesn't necessarily mean you shouldn't do this, but it's worth keeping
> in mind. (I've been burned by this.)
>

This is one of the main reasons why we recommend just importing the module
and referring to objects / variables under there.

https://github.com/secure-systems-lab/code-style-guidelines#imports

Commit:
Add comment to `Signable.__repr__`

The comment stresses the importance of the return value of this
function for backwards compatibility regarding signatures.

Commit:
Remove obsolete line in `Layout.read`

This method used to read a Layout metadata dictionary, containing
a `signatures` field. Now it only reads the layout payload,
contained in a metadata container (next to the signatures field).
Hence the line to get the signatures field has become obsolete
here.

Commit:
Add example argument to Layout.read docstring

lukpueh commented:
Thanks for your comments, @awwad! I addressed some of them directly and ticketized others.
